### PR TITLE
Fix profile image issue on editing username

### PR DESCRIPTION
**Pull request by qazi-fahid**
### Description
Profile image gets broken whenever we edit the username

The issue is fixed by copying the uploaded image files and saving them again for all four sizes according to the new username and then deleting the old images from storage.
The issue which we are facing on qa is fixed. Here is the link to the issue https://rpm.newrelic.com/accounts/631098/applications/4390895/traced_errors/fe236277-5c46-11e9-ada9-0242ac11000a_0_4365

### Testing
- [ ] Run coverage command and mention new % coverage
- [ ] This PR ensures security is maintained w.r.t input validations, output escaping/encoding and privileges against user roles

### Performance
- [ ] Can impact performance (Flag if these changes can affect app performance and add a corresponding JIRA ticket here to conduct performance testing)

### Dependencies
- [ ] This PR is dependent on any other PR or configuration or JIRA issue number; please mention here
- [ ] Includes migration
 
### Post-review
- [x] Rebase and squash commits before merge

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @qazi-fahid! I've created [OSPR-3305](https://openedx.atlassian.net/browse/OSPR-3305) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://open.edx.org/wp-content/uploads/2019/01/individual-contributor-agreement.pdf)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

File: edx_solutions_api_integration/users/views.py [hunk 1]
```diff
@@ -753,6 +753,18 @@ def post(self, request, user_id):  # pylint: disable=R0915
             existing_user.username = username
             response_data['username'] = existing_user.username
             existing_user.save()
+            if old_username != username:
```
Comment (naeem91) [inline edx_solutions_api_integration/users/views.py]: This code is repeating at line 769 too. Did you forget removing the old code?

File: edx_solutions_api_integration/users/views.py [hunk 2]
```diff
@@ -753,6 +753,18 @@ def post(self, request, user_id):  # pylint: disable=R0915
...
+                storage = get_profile_image_storage()
+                profile_image_names = get_profile_image_names(old_username)
+                new_profile_image_names = get_profile_image_names(username)
+                for old_image_size, old_image_name in profile_image_names.items():
+                    if storage.exists(old_image_name):
+                        old_image_path = os.path.join(storage.location, old_image_name)
+                        new_image_path = os.path.join(storage.location, new_profile_image_names[old_image_size])
+                        try:
+                            os.rename(old_image_path, new_image_path)
```
Comment (naeem91) [inline edx_solutions_api_integration/users/views.py]: S3 boto storage does not support renaming files. We need to copy the file to new path and removing the old one.

ref: https://stackoverflow.com/questions/2481685/amazon-s3-boto-how-do-you-rename-a-file-in-a-bucket

File: edx_solutions_api_integration/users/views.py (hunk 1 above)
Comment (qazi-fahid) [reply edx_solutions_api_integration/users/views.py]: Done

File: edx_solutions_api_integration/users/views.py (hunk 2 above)
Comment (qazi-fahid) [reply edx_solutions_api_integration/users/views.py]: Done

File: edx_solutions_api_integration/users/views.py [hunk 3]
```diff
@@ -759,25 +760,16 @@ def post(self, request, user_id):  # pylint: disable=R0915
...
-                storage = get_profile_image_storage()
-                profile_image_names = get_profile_image_names(old_username)
-                new_profile_image_names = get_profile_image_names(username)
-                for old_image_size, old_image_name in profile_image_names.items():
-                    if storage.exists(old_image_name):
-                        old_image_path = storage.path(old_image_name)
-                        new_image_path = os.path.join(storage.location, new_profile_image_names[old_image_size])
-                        try:
-                            os.rename(old_image_path, new_image_path)
+                            file = storage.open(old_image_name)
```
Comment (naeem91) [inline edx_solutions_api_integration/users/views.py]: `file` is a built-in name, please use some other name

File: edx_solutions_api_integration/users/views.py [hunk 4]
```diff
@@ -759,25 +760,16 @@ def post(self, request, user_id):  # pylint: disable=R0915
...
+                            storage.save(new_profile_image_names[old_image_size], ContentFile(file.read()))
                         except OSError:
```
Comment (naeem91) [inline edx_solutions_api_integration/users/views.py]: an `IOError` is raised if there's issue with file opening

File: edx_solutions_api_integration/users/views.py (hunk 3 above)
Comment (qazi-fahid) [reply edx_solutions_api_integration/users/views.py]: done

File: edx_solutions_api_integration/users/views.py (hunk 4 above)
Comment (qazi-fahid) [reply edx_solutions_api_integration/users/views.py]: done

File: edx_solutions_api_integration/users/views.py [hunk 5]
```diff
@@ -761,14 +761,17 @@ def post(self, request, user_id):  # pylint: disable=R0915
                 for old_image_size, old_image_name in profile_image_names.items():
                     if storage.exists(old_image_name):
                         try:
-                            file = storage.open(old_image_name)
-                            storage.save(new_profile_image_names[old_image_size], ContentFile(file.read()))
-                        except OSError:
+                            image_file = storage.open(old_image_name)
+                        except IOError:
```
Comment (naeem91) [inline edx_solutions_api_integration/users/views.py]: the `except` clauses are unnecessary if you are just raising the exceptions. I think you can just enclose the save storage code in try-catch and return an error message

```python
try:
    storage.save(new_profile_image_names[old_image_size], ContentFile(image_file.read()))
except IOError:
    response_data['message'] = _('Could not update profile image')
    return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
else:
   storage.delete(old_image_name)
```

Comment (qazi-fahid) [reply edx_solutions_api_integration/users/views.py]: Done

**Pull request merged**